File: algorithms/2_medium/P/01219_path_with_maximum_gold/backtrack.c
static inline int max_int(const int a, const int b)
{
    return a > b ? a : b;
}

static int dfs(int** grid, const int rows, const int cols, const int i, const int j)
{
    if (i < 0 || j < 0 || i >= rows || j >= cols || grid[i][j] == 0) {
        return 0;
    }

    const int val = grid[i][j];
    int res = 0;

    grid[i][j] = 0;
    res = max_int(res, val + dfs(grid, rows, cols, i - 1, j));
    res = max_int(res, val + dfs(grid, rows, cols, i + 1, j));
    res = max_int(res, val + dfs(grid, rows, cols, i, j - 1));
    res = max_int(res, val + dfs(grid, rows, cols, i, j + 1));
    grid[i][j] = val;

    return res;
}

int getMaximumGold(int** grid, const int rows, const int* cols_data)
{
    assert(rows > 0 && cols_data != NULL && cols_data[0] > 0 && grid != NULL);
    const int cols = cols_data[0];
    assert(rows <= 15 && cols <= 15);

    int res = 0;

    for (int i = 0; i < rows; i += 1) {
        for (int j = 0; j < cols; j += 1) {
            res = max_int(res, dfs(grid, rows, cols, i, j));
        }
    }

    return res;
}

File: algorithms/2_medium/P/01219_path_with_maximum_gold/backtrack.rs
impl Solution {
    fn dfs(grid: &mut [Vec<i32>], cols: isize, i: isize, j: isize) -> i32 {
        let rows = grid.len() as isize;
        if (i < 0 || j < 0 || i >= rows || j >= cols || grid[i as usize][j as usize] == 0) {
            return 0;
        }

        let val = grid[i as usize][j as usize];

        grid[i as usize][j as usize] = 0;
        let res = (val + Self::dfs(grid, cols, i - 1, j))
                  .max(val + Self::dfs(grid, cols, i + 1, j))
                  .max(val + Self::dfs(grid, cols, i, j - 1))
                  .max(val + Self::dfs(grid, cols, i, j + 1));
        grid[i as usize][j as usize] = val;

        res
    }

    pub fn get_maximum_gold(mut grid: Vec<Vec<i32>>) -> i32 {
        assert!(grid.len() > 0 && grid[0].len() > 0);
        let (rows, cols) = (grid.len() as isize, grid[0].len() as isize);
        assert!(rows <= 15 && cols <= 15);

        let mut res = 0_i32;

        for i in 0..rows {
            for j in 0..cols {
                res = res.max(Self::dfs(&mut grid, cols, i, j));
            }
        }

        res
    }
}
